pytorch_lightning/utilities/cli.py

```diff
@@ -153,107 +156,154 @@ def __init__(
         self.datamodule_class = datamodule_class
         self.save_config_callback = save_config_callback
         self.trainer_class = trainer_class
-        self.trainer_defaults = {} if trainer_defaults is None else trainer_defaults
+        self.trainer_defaults = trainer_defaults or {}
+        #self.trainer_fn = trainer_fn
         self.seed_everything_default = seed_everything_default
         self.subclass_mode_model = subclass_mode_model
         self.subclass_mode_data = subclass_mode_data
-        self.parser_kwargs = {} if parser_kwargs is None else parser_kwargs
-        self.parser_kwargs.update({'description': description, 'env_prefix': env_prefix, 'default_env': env_parse})
 
-        self.init_parser()
-        self.add_core_arguments_to_parser()
-        self.add_arguments_to_parser(self.parser)
-        self.parse_arguments()
+        parser_kwargs = parser_kwargs or {}
+        parser_kwargs.update({'description': description, 'env_prefix': env_prefix, 'default_env': env_parse})
+        self.parser = self.init_parser(**parser_kwargs)
+
+        self.add_core_arguments(self.parser)
+        self.add_arguments(self.parser)
+        self.add_subcommands(self.parser)
+        self.parse_arguments(self.parser)
+
         if self.config['seed_everything'] is not None:
-            seed_everything(self.config['seed_everything'])
+            seed_everything(self.config['seed_everything'], workers=True)
+
         self.before_instantiate_classes()
         self.instantiate_classes()
-        self.prepare_fit_kwargs()
-        self.before_fit()
-        self.fit()
-        self.after_fit()
 
-    def init_parser(self) -> None:
+        self.run_subcommand()
+
+    def init_parser(self, use_base: bool = False, **kwargs: Any) -> LightningArgumentParser:
         """Method that instantiates the argument parser"""
-        self.parser = LightningArgumentParser(**self.parser_kwargs)
+        cls = ArgumentParser if use_base else LightningArgumentParser
+        return cls(**kwargs)
 
-    def add_core_arguments_to_parser(self) -> None:
-        """Adds arguments from the core classes to the parser"""
-        self.parser.add_argument(
+    def add_arguments(self, parser: LightningArgumentParser) -> None:
+        """
+        Implement to add extra arguments to the base parser or link arguments
+
+        Args:
+            parser: The base parser object to which arguments can be added
+        """
+        parser.add_argument(
```

Comment:
If the `--seed_everything` argument is moved into the `add_arguments` method that people are supposed to extend, then they are required to do `super().add_arguments(parser)` to not loose `--seed_everything`. I think it is better to not be here.

Reply:
Separating them allows the user to modify seed everything without overriding `add_core_arguments`.
And seed everything is not a "core" argument IMO.

I don't think requiring calling super is asking for too much

Reply:
I still think it should not be in `add_arguments`. If it is the case that people want to remove `seed_everything` then there should be an explicit way to disable it. The way to disable it should not be by omitting a `super()`.

Reply:
Maybe "modifying seed everything" is not a good example.

A better one is in flash, where I am forced to copy-paste the seed everything `add_argument` when I only want to modify the core argument parsing functionality.
(this is because in flash, datamodules are instantiated from class methods)

https://github.com/PyTorchLightning/lightning-flash/blob/1b6de7605ed6353bf7ae8dca3b081e276bdd45a0/flash/utils/cli.py#L53-L58

Reply:
I am not saying that `seed_everything` should be in `add_core_arguments_to_parser`. Just that it shouldn't be in the method that people extend to add arguments. Certainly copy pasting should be avoided. If in `FlashCLI` there is a need copy paste code, then what should be done is to split `LightningCLI.add_core_arguments_to_parser` into more methods so that only the relevant part needs to be modified when extending `FlashCLI`.

Reply:
Okay. Added `add_default_arguments` to add `seed_everything` so `add_arguments` is empty